### PR TITLE
[5.7] Argon and Bcrypt hasher classes do not verify hashing algorithm

### LINKED ISSUE #24162: Argon and Bcrypt hasher classes do not verify hashing algorithm
- Laravel Version: 5.6.20
- PHP Version: 7.2.4
- Database Driver & Version: MySQL 5.7.22

### Description:
`check` method in both `ArgonHasher` & `BcryptHasher` do not verify hashing algorithm. The method only verifies that `$hashedValue` is not an empty string and then proceed to call native php function `password_verify`. Now, since the algorithm and salt are contained in `$hashedValue`, the return value from `password_verify` will not fail if the algorithm used is different than the one intended in the class. 
I had used `bcrypt` function in User model factory in a Laravel project, since the application is still under development; when I upgraded to 5.6, I switched hash driver to argon but forgot to update the factory and later realised I was still able to login with users credentials created with 'broken' factory.

### Steps To Reproduce:
Test method to clarify what I mean:
```php
public function testBasicVerification()
{
    $argonHasher = new \Illuminate\Hashing\ArgonHasher;
    $bcryptHasher = new \Illuminate\Hashing\BcryptHasher;
    $argonHashed = $argonHasher->make('password');
    $bcryptHashed = $bcryptHasher->make('password');
    $this->assertFalse($bcryptHasher->check('password', $argonHashed));
    $this->assertFalse($argonHasher->check('password', $bcryptHashed));
}
```
I'm not sure if you would agree this is a bug and I'm not sure I can think of a 'normal' scenario where this could happen. If I didn't forget to update the factory, I would have never looked at this twice, but on a second thought I think since the method exists in both classes, it makes sense for ArgonHasher to return false when asked to verify a Bcrypt hashed value and vice versa. 

This is a straight forward fix of adding an additional check on `password_get_info($hashedValue)['algo']` value in both methods.

### PR DESCRIPTION
**Pull request by ioluas**
Add additional check on hashing algorithm when checking a hashed value. An exception is thrown if algorithm does not match the one intended in the hasher.

Closes #24162.

### CODE REVIEW

**Comment by tillkruss** [comment]
The tests are failing because PHP 7.2 does not implement the `PASSWORD_AGRON2I` algorithm in it's default build.

We need to either update the Travis CI config, @ioluas maybe you can switch to using the `algoName` instead of `also`?

**Comment by ioluas** [comment]
I just pushed a commit with updated code. That seems to fail too. Maybe I should skip the tests if `PASSWORD_ARGON2I` isn't defined?

**Comment by tillkruss** [comment]
I see, yeah, if you can get Argon working on Travis CI, please do.

**Comment by ioluas** [comment]
Hope that's sorted now. Thanks for your help.

File: src/Illuminate/Hashing/ArgonHasher.php [hunk 1]
```diff
@@ -81,13 +82,19 @@ public function make($value, array $options = [])
      * @param  string  $hashedValue
      * @param  array  $options
      * @return bool
+     *
+     * @throws Exception
```
Comment (a-komarev) [inline src/Illuminate/Hashing/ArgonHasher.php]: Use FQN name here? `@throws \Exception`

File: src/Illuminate/Hashing/BcryptHasher.php [hunk 2]
```diff
@@ -65,13 +66,19 @@ public function make($value, array $options = [])
      * @param  string  $hashedValue
      * @param  array   $options
      * @return bool
+     *
+     * @throws Exception
```
Comment (a-komarev) [inline src/Illuminate/Hashing/BcryptHasher.php]: Use FQN name here? `@throws \Exception`

**Pull request merged**